Add debug logging for process and page state

bnham opened the pull request:
#### 501a16433dc7ebb67f2de223f3dd4fbd2af4fe03
<pre>
Add debug logging for process and page state
<a href="https://bugs.webkit.org/show_bug.cgi?id=245244">https://bugs.webkit.org/show_bug.cgi?id=245244</a>
rdar://problem/99986063

Reviewed by Chris Dumez.

This adds notifyd callbacks that dump state about processes and pages. I&apos;ve found this useful
recently to debug issues with enabling App Nap and adopting RunningBoard on macOS. I also intend to
use this so that Membuster5 knows which domain is loaded into a given WebContent pid. This will
allow the benchmark to provide per-origin memory footprint data.

I also added logging to UserActivity as we&apos;ve needed this logging recently to debug and understand
issues with App Nap activation. As part of this I changed ProcessAssertion to take the description
string as a a string literal to make the logging marginally cheaper.

* Source/WTF/wtf/MemoryPressureHandler.cpp:
(WTF::MemoryPressureHandler::singleton):
(WTF::memoryPressureHandlerIfExists):
(WTF::MemoryPressureHandler::processStateDescription):
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::logMemoryStatistics):
* Source/WebCore/platform/CountedUserActivity.h:
(WebCore::CountedUserActivity::CountedUserActivity):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/UserActivity.cpp:
(WebCore::UserActivity::Impl::Impl):
(WebCore::UserActivity::UserActivity):
* Source/WebCore/platform/UserActivity.h:
* Source/WebCore/platform/mac/UserActivityMac.mm:
(WebCore::UserActivity::Impl::Impl):
(WebCore::UserActivity::Impl::beginActivity):
(WebCore::UserActivity::Impl::endActivity):
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::DisplayCaptureSourceCocoa):
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::AuxiliaryProcess):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::logProcessPoolState):
(WebKit::WebProcessPool::platformInitialize):
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::processAssertionTypeDescription):
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::activityNames):
(WebKit::operator&lt;&lt;):
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::operator&lt;&lt;):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::throttler const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::activityState const):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::additionalStateForDiagnosticReport const):

Canonical link: <a href="https://commits.webkit.org/255239@main">https://commits.webkit.org/255239@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31b812fb24828a996a0922ae68dc057b6dec8a0e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91498 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/621 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101228 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161261 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95503 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/652 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29388 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83844 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97580 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97156 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/414 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78211 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27372 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81999 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70410 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82939 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35628 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16026 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77983 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33412 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17119 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26930 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3629 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39907 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80587 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39133 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36236 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17660 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/MemoryPressureHandler.cpp:
```diff
@@ -248,6 +257,19 @@ void MemoryPressureHandler::setProcessState(WebsamProcessState state)
     m_processState = state;
 }
 
+const char* MemoryPressureHandler::processStateDescription()
```
We prefer ASCIILiteral over `const char*` type for string literals.

cdumez commented on Source/WebCore/page/MemoryRelease.cpp:
```diff
@@ -197,6 +197,7 @@ void logMemoryStatistics(LogMemoryStatisticsReason reason)
     const char* description = logMemoryStatisticsReasonDescription(reason);
 
     RELEASE_LOG(MemoryPressure, "WebKit memory usage statistics at time of %" PUBLIC_LOG_STRING ":", description);
+    RELEASE_LOG(MemoryPressure, "Websam state: %" PUBLIC_LOG_STRING, MemoryPressureHandler::processStateDescription());
```
`MemoryPressureHandler::processStateDescription().characters()` once you switch to ASCIILiteral.

cdumez commented on Source/WebCore/platform/mac/UserActivityMac.mm:
```diff
@@ -26,25 +26,30 @@
 void UserActivity::Impl::beginActivity()
 {
     if (!m_activity) {
+        RELEASE_LOG(ActivityState, "Began activity: %" PUBLIC_LOG_STRING, m_descriptionLiteral.characters());
```
We're not going to have any context in the logging. "Began activity" here is too generic. What kind of activity.
This is why we tend to log the function name, e.g. : `"%p - UserActivity::Impl::beginActivity: description=%" PUBLIC_LOG_STRING`

cdumez replied:
Also, is it really worth keeping the ASCIILiteral as a data member just to avoid calling `[m_description UTF8String]`?

cdumez commented on Source/WebCore/platform/mac/UserActivityMac.mm:
```diff
@@ -26,25 +26,30 @@
 
 void UserActivity::Impl::endActivity()
 {
+    RELEASE_LOG(ActivityState, "Ended activity: %" PUBLIC_LOG_STRING, m_descriptionLiteral.characters());
```
Same comment, logging is too generic.

cdumez commented on Source/WebCore/platform/mac/UserActivityMac.mm:
```diff
@@ -26,25 +26,30 @@
-UserActivity::Impl::Impl(const char* description)
-    : m_description([NSString stringWithUTF8String:description])
+UserActivity::Impl::Impl(ASCIILiteral descriptionLiteral)
+    : m_description([NSString stringWithUTF8String:descriptionLiteral])
```
Should rely on `ASCIILiteral::createNSString()` which is more efficient.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -249,6 +250,20 @@ static AccessibilityPreferences accessibilityPreferences()
+
+        String domain = "unknown"_s;
+        if (process->optionalRegistrableDomain())
+            domain = process->optionalRegistrableDomain()->string();
```
maybe we could do:
```
else
    domain = "unknown"_s;
```
To avoid always constructing a string from "unknown"_s even when there is a registrable domain.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -249,6 +250,20 @@ static AccessibilityPreferences accessibilityPreferences()
+{
+    for (const auto& process : pool.processes()) {
+        WTF::TextStream stream;
+        stream << process;
```
Not sure what TextString does when passing it a `Ref<WebProcessProxy>`. We already print the WebProcessProxy address below, isn't that sufficient?

bnham replied:
Part of the patch is `WTF::TextStream& operator<<(WTF::TextStream&, const WebProcessProxy&)` which prints a bunch of debugging info about each WebProcessProxy to the given stream. For instance:

```
2022-10-04 11:10:24.775686-0700 0x596d0d   Default     0x0                  88125  0    Safari: (WebKit) [com.apple.WebKit:Process] WebProcessProxy 0x120060000 - pid: 88162, pages: 1, visible-pages: 1, has-service-worker, no-assertion-state, domain: nham.me
```

It seems to work since `Ref<T>` defines an implicit conversion to `T&`.

cdumez commented on Source/WebKit/UIProcess/ProcessAssertion.cpp:
```diff
@@ -26,13 +26,32 @@
 
 namespace WebKit {
 
+const char* processAssertionTypeDescription(ProcessAssertionType type)
```
Should return an ASCIILiteral.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -317,4 +317,37 @@ void ProcessThrottler::TimedActivity::updateTimer()
+{
+    Vector<String> names;
+    for (const auto* activity : activities) {
+        if (activity)
```
You may want to skip activities that return true for `isQuietActivity()` too. Those don't have a name and are used for very verbose things like IPC.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -317,4 +317,37 @@ void ProcessThrottler::TimedActivity::updateTimer()
+template <typename T>
+static Vector<String> activityNames(const T& activities)
+{
+    Vector<String> names;
```
Seems like we could reserve capacity here:
`names.reserveInitialCapacity(activities.size());`

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -317,4 +317,37 @@ void ProcessThrottler::TimedActivity::updateTimer()
+    Vector<String> names;
+    for (const auto* activity : activities) {
+        if (activity)
+            names.append(activity->name());
```
If we reversed capacity, we could call `uncheckedAppend()` here.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -317,4 +317,37 @@ void ProcessThrottler::TimedActivity::updateTimer()
+{
+    bool didAppend = false;
+
+    if (auto foregroundActivityNames = activityNames(throttler.m_foregroundActivities); foregroundActivityNames.size()) {
```
Actually, seeing how activityNames() is used, it seems pretty wasteful to allocate a brand new vector. I don't think we should create a new vector just for logging, there is clearly no container mutation going on here.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2137,6 +2138,33 @@ void WebProcessProxy::sendPermissionChanged(WebCore::PermissionName permissionNa
 
+TextStream& operator<<(TextStream& ts, const WebProcessProxy& process)
+{
+    auto appendCount = [&ts](unsigned value, const char *description) {
```
Please use ASCIILiteral for the description

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2137,6 +2138,33 @@ void WebProcessProxy::sendPermissionChanged(WebCore::PermissionName permissionNa
+            ts << ", " << description << ": " << value;
+    };
+
+    auto appendIf = [&ts](bool value, const char *description) {
```
Please use ASCIILiteral for the description

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -663,6 +663,9 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
     bool isVisible() const { return m_activityState.contains(WebCore::ActivityState::IsVisible); }
     bool isVisibleOrOccluded() const { return m_activityState.contains(WebCore::ActivityState::IsVisibleOrOccluded); }
 
+    OptionSet<WebCore::ActivityState::Flag> activityState() { return m_activityState; }
```
Should be marked as `const`.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -491,6 +492,23 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters)
         });
     }
 
+    PAL::registerNotifyCallback("com.apple.WebKit.logPageState"_s, [this] {
```
How is it OK to capture `this` here. What guarantees that `this` is still alive when the notification is received?
I don't see us unregister the callback anyway, I don't even think we currently have an API to unregister. Usually, those notify callbacks are global and iterate over static global containers, they don't capture object pointers like this.

bnham replied:
Should be ok since WebProcess is a NeverDestroyed singleton.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -491,6 +492,23 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters)
 
+    PAL::registerNotifyCallback("com.apple.WebKit.logPageState"_s, [this] {
+        for (auto& page : m_pageMap.values()) {
+            if (!page)
```
Seems impossible to have a null page in there.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -517,6 +518,23 @@ static void initializeXPCConnectionToLogd()
     return auditToken;
 }
 
+const char* WebProcess::processTypeDescription() const
```
Please use ASCIILiteral.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1435,6 +1453,11 @@ void WebProcess::initializeSandbox(const AuxiliaryProcessInitializationParameter
 {
 }
 
+const char* WebProcess::processTypeDescription() const
```
Please use ASCIILiteral

bnham replied:
Deleted this as I wasn't actually using it.

cdumez reviewed (approved):
LGTM

bnham committed:
Add debug logging for process and page state
https://bugs.webkit.org/show_bug.cgi?id=245244
rdar://problem/99986063

Reviewed by Chris Dumez.

This adds notifyd callbacks that dump state about processes and pages. I've found this useful
recently to debug issues with enabling App Nap and adopting RunningBoard on macOS. I also intend to
use this so that Membuster5 knows which domain is loaded into a given WebContent pid. This will
allow the benchmark to provide per-origin memory footprint data.

I also added logging to UserActivity as we've needed this logging recently to debug and understand
issues with App Nap activation. As part of this I changed ProcessAssertion to take the description
string as a a string literal to make the logging marginally cheaper.

* Source/WTF/wtf/MemoryPressureHandler.cpp:
(WTF::MemoryPressureHandler::singleton):
(WTF::memoryPressureHandlerIfExists):
(WTF::MemoryPressureHandler::processStateDescription):
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::logMemoryStatistics):
* Source/WebCore/platform/CountedUserActivity.h:
(WebCore::CountedUserActivity::CountedUserActivity):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/UserActivity.cpp:
(WebCore::UserActivity::Impl::Impl):
(WebCore::UserActivity::UserActivity):
* Source/WebCore/platform/UserActivity.h:
* Source/WebCore/platform/mac/UserActivityMac.mm:
(WebCore::UserActivity::Impl::Impl):
(WebCore::UserActivity::Impl::beginActivity):
(WebCore::UserActivity::Impl::endActivity):
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::DisplayCaptureSourceCocoa):
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::AuxiliaryProcess):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::logProcessPoolState):
(WebKit::WebProcessPool::platformInitialize):
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::processAssertionTypeDescription):
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::activityNames):
(WebKit::operator<<):
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::operator<<):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::throttler const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::activityState const):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::additionalStateForDiagnosticReport const):

Canonical link: https://commits.webkit.org/255239@main
